lib/grandpa/finalisation.go

```diff
@@ -0,0 +1,467 @@
+// Copyright 2022 ChainSafe Systems (ON)
+// SPDX-License-Identifier: LGPL-3.0-only
+
+package grandpa
+
+import (
+	"errors"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/ChainSafe/gossamer/dot/telemetry"
+)
+
+var (
+	errServicesStopFailed       = errors.New("services stop failed")
+	errvotingRoundHandlerFailed = errors.New("voting round ephemeral failed")
+	errfinalisationEngineFailed = errors.New("finalisation engine ephemeral failed")
+)
+
+type ephemeralService interface {
+	Run() error
+	Stop() error
+}
+
+type finalisationHandler struct {
+	servicesLock       sync.Mutex
+	finalisationEngine ephemeralService
+	votingRound        ephemeralService
+
+	// newServices is a constructor function which takes care to instantiate
+	// and return the services needed to finalize a round, those services
+	// are ephemeral services with a lifetime of a round
+	newServices   func() (engine, voting ephemeralService)
+	initiateRound func() error
+
+	stopCh      chan struct{}
+	handlerDone chan struct{}
+}
+
+func newFinalisationHandler(service *Service) *finalisationHandler {
+	return &finalisationHandler{
+		newServices: func() (engine, voting ephemeralService) {
+			finalisationEngine := newfinalisationEngine(service)
+			votingRound := newvotingRoundHandler(service, finalisationEngine.actionCh)
```

Comment:
Im still getting familiar with best practices for service orchestration, so had a question.

Why do message passing directly between two subchannels, rather than have the handlerService be in charge of message propagation?

My initial though its because its more efficient/easier to do it directly, but does it add complication to graceful shutdown if one of the sub services fails?

Reply:
So the message exchange is between `finalization engine (ephemeral) -> voting round handler (ephemeral)`, it is not worth it for `finalization handler` to know about those messages because it does not care about them it just cares about keeping ephemeral services running for one entire round. So, given that `finalization handler` just manages the ephemeral services it is easy to manage the shutdown because whenever one ephemeral service fails (`finalization engine` or `voting round handler`) the manager just stops the other, but if the handler knows that both ephemeral services ended without error it proceeds to the next round.

Reply:
okay sweet, makes sense. Thank you!